primer-miso/src/Primer/Miso/Util.hs

```diff
@@ -328,21 +364,41 @@ availableForSelection ::
   DefMap ->
   Level ->
   Editable ->
-  ASTDef ->
-  DefSelection a ->
+  Either (ASTTypeDef TypeMeta KindMeta) ASTDef ->
+  Selection' a ->
   [Action]
-availableForSelection tydefs defs level editable def defSel = case defSel.node of
-  Nothing -> Available.forDef defs level editable defSel.def
-  Just nodeSel -> case nodeSel.nodeType of
-    BodyNode -> Available.forBody tydefs level editable (astDefExpr def) (getID nodeSel)
-    SigNode -> Available.forSig level editable (astDefType def) (getID nodeSel)
+availableForSelection tydefs defs level editable defOrTypeDef sel = case (sel, defOrTypeDef) of
+  (SelectionTypeDef TypeDefSelection{def = defName, node}, Left def) -> case node of
+    Just (TypeDefParamNodeSelection sel') -> case sel'.kindMeta of
+      Nothing -> Available.forTypeDefParamNode sel'.param level editable tydefs defs defName def
+      Just sel'' -> Available.forTypeDefParamKindNode sel'.param (getID sel'') level editable tydefs defs defName def
+    Just (TypeDefConsNodeSelection sel') -> case sel'.field of
+      Nothing -> Available.forTypeDefConsNode level editable tydefs defs defName def
+      Just sel'' -> Available.forTypeDefConsFieldNode sel'.con sel''.index (getID sel''.meta) level editable tydefs defs defName def
+    Nothing -> Available.forTypeDef level editable tydefs defs defName def
+  (SelectionDef DefSelection{def = defName, node}, Right def) -> case node of
+    Nothing -> Available.forDef defs level editable defName
+    Just nodeSel -> case nodeSel.nodeType of
+      BodyNode -> Available.forBody tydefs level editable (astDefExpr def) (getID nodeSel)
+      SigNode -> Available.forSig level editable (astDefType def) (getID nodeSel)
+  _ -> [] -- TODO allow raising a warning instead (when def and selection don't match)
```

Comment:
Oof, any thoughts on how to make this block of code less verbose/repetitive?

Reply:
Actually, as alluded to in the comment above this function, this arguably should be the interface for that module, i.e. just take a selection.